### PR TITLE
derp/derpmap: new DERP config package, merge netcheck into magicsock

### LINKED ISSUE #153: DERP in magicsock should not use global state
The DERP logic in magicsock uses global variables to track existing DERP servers. This makes it hard to test (need some swaps and cleanups) and impossible to parallelize any test that uses or cares about DERP.

All the DERP glue in magicsock should be a standalone struct that can be initialized to defaults, or overridden in tests via magicsock Options.

### PR DESCRIPTION
**Pull request by bradfitz**
Fixes #153
Updates #162
Updates #163

Signed-off-by: Brad Fitzpatrick <bradfitz@tailscale.com>

### CODE REVIEW

**Comment by bradfitz** [comment]
FYI only. Still WIP. Just uploading before end of day here.

File: stunner/stunner.go [hunk 1]
```diff
@@ -131,6 +131,10 @@ func (s *Stunner) resolver() *net.Resolver {
 func (s *Stunner) Run(ctx context.Context) error {
 	s.sessions = map[string]*session{}
 	for _, server := range s.Servers {
+		if _, _, err := net.SplitHostPort(server); err != nil {
+			panic("bogus")
```
Comment (crawshaw) [inline stunner/stunner.go]: remnant

File: derp/derpmap/derpmap.go [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+	stun6   []string
+}
+
+func (w *World) IDs() []int                { return w.ids }
+func (w *World) STUN4() []string           { return w.stun4 }
+func (w *World) STUN6() []string           { return w.stun6 }
+func (w *World) ServerByID(id int) *Server { return w.byID[id] }
+
+// LocationOfID returns the geographic name of a node, if present.
+func (w *World) LocationOfID(id int) string {
```
[the remark this answers is not in the document]
Comment (bradfitz) [reply derp/derpmap/derpmap.go]: Let's wait until it's needed. I can only think of contrived uses at the moment.

Easy to add later.

**Comment by bradfitz** [comment]
The magicsock sock tests still seem flaky. I do see it sometimes pass, but often it fails. It seems more reliable before this commit, though, so perhaps I made it worse somehow. Or just exacerbated an existing race.

Any ideas? Can you try to patch this in and repro, @crawshaw?

**Comment by bradfitz** [comment]
Found the flake reason: stunner was starting a reading goroutine (competing for packets) even when c.GetSTUNConn4 was non-nil.

**Comment by bradfitz** [comment]
Okay, I'm happy with this. Tests seem happy & manual testing seems happy.

**Pull request merged**